hpx/lcos/async.hpp

```diff
@@ -65,8 +90,18 @@ namespace hpx { namespace detail
             static_assert(is_valid::value,
                 "The action to invoke is not supported by the target");
 
-            return hpx::detail::async_impl<Action>(launch_policy, c.get_id(),
-                std::forward<Ts>(ts)...);
+            // invoke directly if client is ready
+            if (c.is_ready())
```

Comment:
What I meant is to have this optimization directly in client_base. I was under the impression we had the same we have for `future<T>::then` ... the only optimization we in that regard is here: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/lcos/detail/future_data.hpp#L565

I missed the part that we need to create a new future inside ::then ... sorry for the noise.